File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int beginOfSubstring = 0;
        Map<Character, Integer> lastPositionOfChar = new HashMap<>();
        int result = 0;
        for(int i=0; i<s.length(); i++){
            if(lastPositionOfChar.containsKey(s.charAt(i))&&lastPositionOfChar.get(s.charAt(i)) >= beginOfSubstring){
                beginOfSubstring = lastPositionOfChar.get(s.charAt(i))+1;
            }
            result = Math.max(result, i-beginOfSubstring+1);
            lastPositionOfChar.put(s.charAt(i), i);
        }
        return result;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        begin_of_substring = 0
        last_position_of_char = {}
        result = 0

        for i, ch in enumerate(s):
            if ch in last_position_of_char and last_position_of_char[ch] >= begin_of_substring:
                begin_of_substring = last_position_of_char[ch] + 1
            result = max(result, i - begin_of_substring + 1)
            last_position_of_char[ch] = i

        return result
